pkg/virt-launcher/virtwrap/sriov/hostdev.go

```diff
@@ -98,3 +102,88 @@ func decoratePciAddressField(addressField string) (*api.Address, error) {
 	}
 	return decoratedAddrField, nil
 }
+
+type deviceDetacher interface {
+	DetachDevice(xmlData string) error
+}
+
+func DetachHostDevices(dom deviceDetacher, hostDevices []api.HostDevice) error {
+	for _, device := range hostDevices {
+		devXML, err := xml.Marshal(device)
+		if err != nil {
+			return fmt.Errorf("failed to encode (xml) device %v, err: %v", device, err)
+		}
+		err = dom.DetachDevice(string(devXML))
+		if err != nil {
+			return fmt.Errorf("failed to detach device %s, err: %v", devXML, err)
+		}
+		log.Log.Infof("Successfully hot-unplug device: %s (%v)", device.Alias.Name, device.Source.Address)
+	}
+	return nil
+}
+
+func FilterHostDevices(domainSpec *api.DomainSpec) []api.HostDevice {
+	var hostDevices []api.HostDevice
+
+	for _, device := range domainSpec.Devices.HostDevices {
+		if device.Alias != nil && strings.HasPrefix(device.Alias.Name, AliasPrefix) {
+			hostDevices = append(hostDevices, device)
+		}
+	}
+	return hostDevices
+}
+
+type Registrar interface {
+	Register(callback interface{}) error
+	Deregister() error
+}
+
+func WaitHostDevicesToDetach(detacher Registrar, hostDevices []api.HostDevice, timeout time.Duration, f func() error) error {
+	if len(hostDevices) == 0 {
+		log.Log.Info("No host-devices to detach.")
+		return nil
+	}
+
+	eventsChan := make(chan string, len(hostDevices)+2)
+	var callback libvirt.DomainEventDeviceRemovedCallback = func(c *libvirt.Connect, d *libvirt.Domain, event *libvirt.DomainEventDeviceRemoved) {
+		eventsChan <- event.DevAlias
+	}
+
+	if err := detacher.Register(callback); err != nil {
+		return fmt.Errorf("failed to wait for host-devices detach: %v", err)
+	}
+	defer func() {
+		if err := detacher.Deregister(); err != nil {
+			log.Log.Reason(err).Errorf("failed to wait for host-devices detach, deregister failure: %v", err)
+		}
+	}()
+
+	if err := f(); err != nil {
+		return err
+	}
+
+	var detachedHostDevices []string
+	for {
+		select {
+		case deviceAlias := <-eventsChan:
+			if dev := deviceLookup(hostDevices, deviceAlias); dev != nil {
+				detachedHostDevices = append(detachedHostDevices, dev.Alias.Name)
+			}
+			if len(hostDevices) == len(detachedHostDevices) {
+				return nil
+			}
+		case <-time.After(timeout):
+			return fmt.Errorf("failed to wait for host-devices detach, timeout reached: %v", detachedHostDevices)
+		}
```

Comment:
Instead of throwing an error directly maybe we can programtically check if the devices were correctly detached and if not throw and error (don't know if we can do so)? maybe we reach this code after libvirt missing some event or the like.

Also in case we can check for them to be detaching polling somewhere for state we can refactor this and remove the channel and callbacks.